shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API/Validation.hs

```diff
@@ -40,13 +51,269 @@ import qualified Shelley.Spec.Ledger.STS.Bbody as STS
+      , [ApplyBlockData era]
+      )
+
+
+data ApplyBlockData era
+  = ApplyTick (Signal (Core.EraRule "TICK" era))
+  | ApplyBlock (Signal (Core.EraRule "BBODY" era))
+
+deriving instance
+  ( Eq (Signal (Core.EraRule "TICK" era))
+  , Eq (Signal (Core.EraRule "BBODY" era))
+  ) => Eq (ApplyBlockData era)
+
+deriving instance
+  ( Ord (Signal (Core.EraRule "TICK" era))
+  , Ord (Signal (Core.EraRule "BBODY" era))
+  ) => Ord (ApplyBlockData era)
+
+deriving instance
+  ( Show (Signal (Core.EraRule "TICK" era))
+  , Show (Signal (Core.EraRule "BBODY" era))
+  ) => Show (ApplyBlockData era)
+
+data ApplyBlockTransitionError era
+   = ApplyBlockTransitionError_Block (BlockTransitionError era)
+   | ApplyBlockTransitionError_Tick (TickTransitionError era)
+
+deriving instance
+ ( Show (BlockTransitionError era)
+ , Show (TickTransitionError era)
+ ) => Show (ApplyBlockTransitionError era)
+
+deriving instance
+ ( Eq (BlockTransitionError era)
+ , Eq (TickTransitionError era)
+ ) => Eq (ApplyBlockTransitionError era)
+
+deriving instance
+ ( Ord (BlockTransitionError era)
+ , Ord (TickTransitionError era)
+ ) => Ord (ApplyBlockTransitionError era)
+
+type ApplyBlockError era =
+  ( ApplyBlockTransitionError era
+  , State (Core.EraRule "TICK" era)
+  , ApplyBlockData era
+  )
+
+runApplyBlockData
+  :: forall era. ApplyBlock' era
+  => ApplySTSOpts
+  -> State (Core.EraRule "TICK" era)
+  -> [ApplyBlockData era]
+  -> BaseM (Core.EraRule "TICK" era)
+    (Either (ApplyBlockError era) (State (Core.EraRule "TICK" era)))
+runApplyBlockData opts = go
+  where
+    loop
+      :: ([a] -> (ApplyBlockTransitionError era))
+      -> ApplyBlockData era
+      -> [ApplyBlockData era]
+      -> (State (Core.EraRule "TICK" era), [[a]])
+      -> BaseM (Core.EraRule "TICK" era)
+        (Either (ApplyBlockError era) (State (Core.EraRule "TICK" era)))
+    loop _ _ xs (good, []) = go good xs
+    loop c x _ (st, bad) = pure . Left . (,st,x) . c $ join bad
+    {-# INLINE loop #-}
+
+    go
+      :: State (Core.EraRule "TICK" era)
+      -> [ApplyBlockData era]
+      -> BaseM (Core.EraRule "TICK" era)
+        (Either (ApplyBlockError era) (State (Core.EraRule "TICK" era)))
+    go st [] = pure (Right st)
+    go st0 (x:xs) = do
+      case x of
+        ApplyTick tick -> applyTickOpts' opts (Proxy :: Proxy era) st0 tick
+          >>= loop (ApplyBlockTransitionError_Tick . TickTransitionError) x xs
+        ApplyBlock blk -> applyBlockOpts' opts (Proxy :: Proxy era) st0 blk
+          >>= loop (ApplyBlockTransitionError_Block . BlockTransitionError) x xs
+
+-- TODO: move implementations out of class
+-- TODO: move defaulting methods to a suitable "deriving via" newtype
 class
-  ( ChainData (Block era),
+    ( STS (Core.EraRule "TICK" era)
+    , STS (Core.EraRule "BBODY" era)
+    , Environment (Core.EraRule "TICK" era) ~ ()
+    , BaseM (Core.EraRule "TICK" era) ~ BaseM (Core.EraRule "BBODY" era)
+    ) => ApplyBlock' era where
```

Comment:
I still don't understand the point of this, or why it's needed. We should be testing the existing API

Reply:
I've combined the two classes and specialised these types to ShelleyBase, NewEraState, SlotNo, and Block; as appropriate.